template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -243,7 +250,39 @@ private Map<String, Set<ServiceComponent>> getNonGatewayServicesByHostGroup() {
                                 .filter(i -> !i.getComponent().equalsIgnoreCase("GATEWAY"))
                                 .collect(toSet())
                 ));
+    }
+
+    private Set<String> collectComponentByHostGroupWithRoleDependencies() {
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        return hgToNonGwServiceComponents.values().stream()
+                .filter(serviceComponents -> !Strings.isNullOrEmpty(getDependentService(serviceComponents)))
+                .map(serviceComponents -> ROLE_DEPENDENCY.get(getDependentService(serviceComponents)))
+                .collect(toSet());
+    }
+
+    private String getDependentService(Set<ServiceComponent> serviceComponents) {
+        Set<String> serviceComponentRoles = collectComponents(serviceComponents);
+        return ROLE_DEPENDENCY.keySet().stream().filter(serviceComponentRoles::contains).findFirst().orElse(null);
+    }
+
+    public List<String> getUnhealthyDependentHostGroups(StackDto stackdto) {
```

Comment:
Add JavaDoc. If the depdenent hostgroups cannot be identified - this method should return either a special value, or a list of all host groups. Some way of saying "NotDetermined" will work better - since actions fallback operations can be performed as a result of this.